File: flowersec-ts/src/v4/recordReference.test.ts
// Fixed public fixtures; no runtime READY, sequence/replay or key-use authority.
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import { chacha20poly1305 } from "@noble/ciphers/chacha.js";
import { gcm } from "@noble/ciphers/aes.js";
import { expand } from "@noble/hashes/hkdf.js";
import { sha256 } from "@noble/hashes/sha2.js";
import { transportV4Domains, transportV4RecordRegistry, transportV4Registry } from "../generated/transportV4Registry.js";

type Field = { name: string; type: string; const?: number; max?: number };
type Part = { name: string; encoding: string; length?: number; enum?: readonly number[] };
type Domain = { name: string; label_bytes: string; input_schema: { parts: Part[] } };
type Vector = {
  id: string; profile: string; algorithm: string; epoch: number; direction: number;
  sequence_scope: string; sequence: string; frame_type: number; epoch_root_hex: string; handshake_hash_hex: string;
  key_info_hex: string; key_hex: string; nonce_hex: string; envelope_header_hex: string; record_header_hex: string;
  aad_hex: string; plaintext_hex: string; ciphertext_hex: string; wire_hex: string;
};
type Negative = { id: string; source: string; field: "key" | "nonce" | "aad" | "ciphertext"; value_hex: string; expected_error: string };
const corpus = JSON.parse(readFileSync(new URL("../../../testdata/transport_v4/records.json", import.meta.url), "utf8")) as {
  schema_sha256: string; vectors: Vector[]; negatives: Negative[];
};
const registry = transportV4RecordRegistry as { header: readonly Field[]; nonce: readonly Field[]; envelope: { layout: readonly Field[] }; profiles: Record<string, { record_aead: string; tag_bytes: number }> };
const domains = transportV4Domains as unknown as Domain[];
const bytes = (hex: string) => Uint8Array.from(Buffer.from(hex, "hex"));
const join = (...values: Uint8Array[]) => Uint8Array.from(Buffer.concat(values));
const text = (value: string) => new TextEncoder().encode(value);

function layout(fields: readonly Field[], values: Record<string, number | bigint>): Uint8Array {
  return join(...fields.map(field => {
    const width = ({ uint8: 1, uint16_be: 2, uint32_be: 4, uint64_be: 8 } as Record<string, number>)[field.type];
    const supplied = field.const ?? values[field.name];
    if (!width || supplied === undefined || (typeof supplied === "number" && !Number.isSafeInteger(supplied))) throw new Error("invalid record integer");
    const value = BigInt(supplied);
    if (value < 0n || value >= 1n << BigInt(width * 8) || (field.max !== undefined && value > BigInt(field.max))) throw new Error("record integer range");
    const out = new Uint8Array(8);
    new DataView(out.buffer).setBigUint64(0, value, false);
    return out.slice(8 - width);
  }));
}

function domain(name: string, values: Record<string, Uint8Array>, integers: Record<string, number | bigint>): Uint8Array {
  const spec = domains.find(d => d.name === name);
  if (!spec) throw new Error("missing domain");
  return join(bytes(spec.label_bytes), ...spec.input_schema.parts.map(part => {
    if (["raw", "lp-bytes", "lp-ascii"].includes(part.encoding)) {
      const value = values[part.name];
      if (!value || (part.length !== undefined && part.length !== value.length)) throw new Error("invalid domain bytes");
      if (part.encoding === "lp-ascii" && value.some(b => b > 127)) throw new Error("non-ascii profile");
      return part.encoding === "raw" ? value : join(layout([{ name: "length", type: "uint32_be" }], { length: value.length }), value);
    }
    const type = ({ u8: "uint8", u32: "uint32_be", u64: "uint64_be" } as Record<string, string>)[part.encoding];
    if (!type || (part.enum && !part.enum.some(n => BigInt(n) === BigInt(integers[part.name]!)))) throw new Error("invalid domain integer");
    return layout([{ name: part.name, type }], integers);
  }));
}

function cipher(algorithm: string, key: Uint8Array, nonce: Uint8Array, aad: Uint8Array) {
  if (key.length !== 32) throw new Error("record key length");
  switch (algorithm) {
    case "chacha20-poly1305": return chacha20poly1305(key, nonce, aad);
    case "aes-256-gcm": return gcm(key, nonce, aad);
    default: throw new Error("unknown record algorithm");
  }
}

describe("v4.record.ts_reference", () => {
  it("compares independent encodings, key derivation and both AEAD profiles", () => {
    expect(corpus.schema_sha256).toBe(transportV4Registry.schemaSHA256);
    expect(corpus.vectors.length).toBeGreaterThan(0);
    for (const v of corpus.vectors) {
      const spec = registry.profiles[v.profile]!;
      expect(spec.record_aead).toBe(v.algorithm);
      const integers = { epoch: v.epoch, direction: v.direction, sequence_scope: BigInt(v.sequence_scope), sequence: BigInt(v.sequence) };
      const header = layout(registry.header, integers), nonce = layout(registry.nonce, integers), plaintext = bytes(v.plaintext_hex);
      const envelope = layout(registry.envelope.layout, { payload_length: header.length + plaintext.length + spec.tag_bytes, frame_type: v.frame_type });
      const info = domain("record_key", { profile: text(v.profile), handshake_hash: bytes(v.handshake_hash_hex) }, integers);
      const key = expand(sha256, bytes(v.epoch_root_hex), info, 32);
      const aad = domain("record_aad", { profile: text(v.profile), envelope_header: envelope, record_header: header }, integers);
      const ciphertext = cipher(spec.record_aead, key, nonce, aad).encrypt(plaintext);
      for (const [actual, expected] of [[header, v.record_header_hex], [nonce, v.nonce_hex], [envelope, v.envelope_header_hex],
        [info, v.key_info_hex], [key, v.key_hex], [aad, v.aad_hex], [ciphertext, v.ciphertext_hex],
        [join(envelope, header, ciphertext), v.wire_hex]] as const) expect(actual, v.id).toEqual(bytes(expected));
      expect(cipher(spec.record_aead, key, nonce, aad).decrypt(ciphertext), v.id).toEqual(plaintext);
    }
  });

  it("rejects every generated authentication mutation", () => {
    expect(corpus.negatives.length).toBeGreaterThan(0);
    const byID = new Map(corpus.vectors.map(v => [v.id, v]));
    for (const n of corpus.negatives) {
      expect(n.expected_error).toBe("record_authentication_failed");
      expect(["key", "nonce", "aad", "ciphertext"]).toContain(n.field);
      const original = byID.get(n.source);
      expect(original).toBeDefined();
      const v = { ...original!, [`${n.field}_hex`]: n.value_hex };
      expect(() => cipher(v.algorithm, bytes(v.key_hex), bytes(v.nonce_hex), bytes(v.aad_hex)).decrypt(bytes(v.ciphertext_hex)), n.id).toThrow();
    }
  });
});

File: flowersec-go/internal/protocolv4/record_reference_test.go
package protocolv4

// Fixed public record fixtures only. These tests do not establish READY,
// authorize epochs/scopes/sequences or replace live key-use and owner gates.
import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hkdf"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"testing"

	"golang.org/x/crypto/chacha20poly1305"
)

type recordField struct {
	Name, Type string
	Const, Max *uint64
}

func recordLayout(fields []recordField, values map[string]uint64) ([]byte, error) {
	var out []byte
	for _, field := range fields {
		value, exists := values[field.Name]
		if field.Const != nil {
			value, exists = *field.Const, true
		}
		width := map[string]int{"uint8": 1, "uint16_be": 2, "uint32_be": 4, "uint64_be": 8}[field.Type]
		if !exists || width == 0 || (width < 8 && value >= uint64(1)<<(width*8)) || (field.Max != nil && value > *field.Max) {
			return nil, fmt.Errorf("invalid record integer %s", field.Name)
		}
		encoded := binary.BigEndian.AppendUint64(nil, value)
		out = append(out, encoded[8-width:]...)
	}
	return out, nil
}

func recordDomain(name string, byteValues map[string][]byte, integerValues map[string]uint64) ([]byte, error) {
	var domains []struct {
		Name        string
		Label       string `json:"label_bytes"`
		InputSchema struct {
			Parts []struct {
				Name, Encoding string
				Length         int
				Enum           []uint64
			} `json:"parts"`
		} `json:"input_schema"`
	}
	if err := json.Unmarshal([]byte(DomainRegistryJSON), &domains); err != nil {
		return nil, err
	}
	for _, domain := range domains {
		if domain.Name != name {
			continue
		}
		out, err := hex.DecodeString(domain.Label)
		if err != nil {
			return nil, err
		}
		for _, part := range domain.InputSchema.Parts {
			switch part.Encoding {
			case "raw", "lp-bytes", "lp-ascii", "lp-map":
				value, exists := byteValues[part.Name]
				if !exists || (part.Length != 0 && len(value) != part.Length) || uint64(len(value)) > uint64(^uint32(0)) {
					return nil, fmt.Errorf("invalid record domain bytes %s", part.Name)
				}
				if part.Encoding == "lp-ascii" {
					for _, b := range value {
						if b >= 128 {
							return nil, fmt.Errorf("non-ascii profile")
						}
					}
				}
				if part.Encoding != "raw" {
					out = binary.BigEndian.AppendUint32(out, uint32(len(value)))
				}
				out = append(out, value...)
			case "u8", "u32", "u64":
				if len(part.Enum) > 0 {
					matched := false
					for _, value := range part.Enum {
						matched = matched || value == integerValues[part.Name]
					}
					if !matched {
						return nil, fmt.Errorf("invalid record enum")
					}
				}
				typeName := map[string]string{"u8": "uint8", "u32": "uint32_be", "u64": "uint64_be"}[part.Encoding]
				encoded, err := recordLayout([]recordField{{Name: part.Name, Type: typeName}}, integerValues)
				if err != nil {
					return nil, err
				}
				out = append(out, encoded...)
			default:
				return nil, fmt.Errorf("unsupported record domain field")
			}
		}
		return out, nil
	}
	return nil, fmt.Errorf("record domain missing")
}

type recordVector struct {
	ID, Profile, Algorithm string
	Epoch, Direction       uint64
	Scope                  uint64 `json:"sequence_scope,string"`
	Sequence               uint64 `json:"sequence,string"`
	FrameType              uint64 `json:"frame_type"`
	Root                   string `json:"epoch_root_hex"`
	Hash                   string `json:"handshake_hash_hex"`
	KeyInfo                string `json:"key_info_hex"`
	Key                    string `json:"key_hex"`
	Nonce                  string `json:"nonce_hex"`
	Envelope               string `json:"envelope_header_hex"`
	Header                 string `json:"record_header_hex"`
	AAD                    string `json:"aad_hex"`
	Plaintext              string `json:"plaintext_hex"`
	Ciphertext             string `json:"ciphertext_hex"`
	Wire                   string `json:"wire_hex"`
}

func recordAEAD(algorithm string, key []byte) (cipher.AEAD, error) {
	if len(key) != 32 {
		return nil, fmt.Errorf("invalid record key length")
	}
	switch algorithm {
	case "chacha20-poly1305":
		return chacha20poly1305.New(key)
	case "aes-256-gcm":
		block, err := aes.NewCipher(key)
		if err != nil {
			return nil, err
		}
		return cipher.NewGCM(block)
	default:
		return nil, fmt.Errorf("unsupported record algorithm")
	}
}

func TestV4RecordSharedCorpus(t *testing.T) {
	var corpus struct {
		Schema    string `json:"schema_sha256"`
		Vectors   []recordVector
		Negatives []struct {
			ID, Source, Field string
			Value             string `json:"value_hex"`
			Error             string `json:"expected_error"`
		}
	}
	raw, err := os.ReadFile("../../../testdata/transport_v4/records.json")
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal(raw, &corpus); err != nil {
		t.Fatal(err)
	}
	if corpus.Schema != SchemaSHA256 || len(corpus.Vectors) == 0 || len(corpus.Negatives) == 0 {
		t.Fatal("record corpus binding/coverage")
	}
	var registry struct {
		Envelope      struct{ Layout []recordField }
		Header, Nonce []recordField
		Profiles      map[string]struct {
			Algorithm string `json:"record_aead"`
			TagBytes  uint64 `json:"tag_bytes"`
		}
	}
	if err := json.Unmarshal([]byte(RecordRegistryJSON), &registry); err != nil {
		t.Fatal(err)
	}
	byID := map[string]recordVector{}
	for _, v := range corpus.Vectors {
		byID[v.ID] = v
		profile, exists := registry.Profiles[v.Profile]
		if !exists || profile.Algorithm != v.Algorithm {
			t.Fatal("record profile drift", v.ID)
		}
		integers := map[string]uint64{"epoch": v.Epoch, "sequence_scope": v.Scope, "sequence": v.Sequence, "direction": v.Direction}
		header, err := recordLayout(registry.Header, integers)
		if err != nil {
			t.Fatal(err)
		}
		nonce, err := recordLayout(registry.Nonce, integers)
		if err != nil {
			t.Fatal(err)
		}
		plain := noiseHex(t, v.Plaintext)
		envelope, err := recordLayout(registry.Envelope.Layout, map[string]uint64{
			"payload_length": uint64(len(header)+len(plain)) + profile.TagBytes, "frame_type": v.FrameType,
		})
		if err != nil {
			t.Fatal(err)
		}
		info, err := recordDomain("record_key", map[string][]byte{"profile": []byte(v.Profile), "handshake_hash": noiseHex(t, v.Hash)}, integers)
		if err != nil {
			t.Fatal(err)
		}
		key, err := hkdf.Expand(sha256.New, noiseHex(t, v.Root), string(info), 32)
		if err != nil {
			t.Fatal(err)
		}
		aad, err := recordDomain("record_aad", map[string][]byte{"profile": []byte(v.Profile), "envelope_header": envelope, "record_header": header}, integers)
		if err != nil {
			t.Fatal(err)
		}
		aead, err := recordAEAD(profile.Algorithm, key)
		if err != nil {
			t.Fatal(err)
		}
		ciphertext := aead.Seal(nil, nonce, plain, aad)
		wire := bytes.Join([][]byte{envelope, header, ciphertext}, nil)
		for _, pair := range []struct {
			actual   []byte
			expected string
		}{
			{header, v.Header}, {nonce, v.Nonce}, {envelope, v.Envelope}, {info, v.KeyInfo}, {key, v.Key}, {aad, v.AAD}, {ciphertext, v.Ciphertext}, {wire, v.Wire},
		} {
			if hex.EncodeToString(pair.actual) != pair.expected {
				t.Fatal("record bytes differ", v.ID)
			}
		}
		opened, err := aead.Open(nil, nonce, ciphertext, aad)
		if err != nil || !bytes.Equal(opened, plain) {
			t.Fatal("record open differs", v.ID, err)
		}
	}
	for _, n := range corpus.Negatives {
		v, exists := byID[n.Source]
		if !exists || n.Error != "record_authentication_failed" {
			t.Fatal("invalid negative", n.ID)
		}
		key, nonce, aad, ciphertext := noiseHex(t, v.Key), noiseHex(t, v.Nonce), noiseHex(t, v.AAD), noiseHex(t, v.Ciphertext)
		switch n.Field {
		case "key":
			key = noiseHex(t, n.Value)
		case "nonce":
			nonce = noiseHex(t, n.Value)
		case "aad":
			aad = noiseHex(t, n.Value)
		case "ciphertext":
			ciphertext = noiseHex(t, n.Value)
		default:
			t.Fatal("unknown negative field", n.ID)
		}
		aead, err := recordAEAD(v.Algorithm, key)
		if err != nil {
			t.Fatal(err)
		}
		if plain, err := aead.Open(nil, nonce, ciphertext, aad); err == nil || len(plain) != 0 {
			t.Fatal("unauthenticated plaintext", n.ID)
		}
	}
}
